Monday muddiest points notebook (#29)
AlexsLemonade/2022-september-training: supplemental_notebooks/2022-09-19_muddiest_points.Rmd

1. Click the `Code` button in the upper right of this page and choose `Download Rmd` to save this notebook (`2022-09-20_muddiest_points.Rmd`) to your computer.
If you have RStudio installed on your computer, you can then open the file directly.

Otherwise you can add it to the CCDL RStudio server to work with it there with the following additional steps:

2. Navigate to RStudio server.  
3. In the `File` pane, navigate to `training-modules/intro-to-R-tidyverse`.     
4. Click the `Upload` button in the `File` pane.   
5. Click `Choose file` and find this `Rmd` file on your computer, then click `OK`  

Now you can open up this file and run it on the RStudio Server. 

## RDS and RData formats

As we discussed in the Introduction to Tidyverse lesson, there is an R-specific file format called "RDS"  (R Data Serialized/Single). 
This file format will store data exactly as you have it in your R environment, making it easy to export and load in any kind of R object (aka variable). 
In addition, RDS files can be compressed, meaning they will take up less space on your computer.

> In base R, RDS files can be saved with the function `saveRDS()` and read in with `readRDS`.

Reviewer: Formatting consistency
```suggestion
In base R, RDS files can be saved with the function `saveRDS()` and read in with `readRDS()`.
```

There are complementary functions in the `readr` package, `readr::write_rds()` and `readr::read_rds()`, that can be used instead.

> The main difference between base R and tidyverse versions of these functions are that, when writing to a file, `saveRDS()` will compress the data by default but `readr::write_rds()` will not unless you explicitly specificy with the `compress` argument (see `?readr::write_rds()` for help!).

Reviewer: ```suggestion
The main difference between base R and tidyverse versions of these functions is that, when writing to a file, `saveRDS()` will compress the data by default but `readr::write_rds()` will not unless you explicitly specify a compression method with the `compress` argument (see `?readr::write_rds()` for help!).
```

 You can also use the R command `ls()` to see all defined variables in the current session.
- Determine which objects you want to save as RDS files, and export _each separately_ as an RDS file, with either `saveRDS()` or `readr::write_rds()`.

## Labeling individual points in plots

Often, we wish to label points in plots, for example labeling individual points (e.g. points representing genes of interest) in a volcano plot. 
The example code below shows you how to include labeling for a subset of points of interest.

First, let's create a toy dataset to plot with.
Here, I've taken a random subset of 10 "male_female" contrast rows from the data frame we plotted in Intro ggplot.

```{r}
# Create a small dataset to use here as a plotting example. This data is actually
# a random selection of 10 "male_female" contrast rows from the data frame we 
# plotted in Intro ggplot.
# The tribble() function from the tibble package (tibble::tribble()) will create a
#  tibble, which _is_ a data frame but with a couple helpful tidyverse features
#  More info here: https://tibble.tidyverse.org/articles/tibble.html

> gene_data <- tibble::tribble(

Reviewer: I don't _love_ introducing `tribble` here. I tend to think reinforcing the core features is better at this point than adding new syntax, so I would use a base `data.frame` here.

    y = neg_log10_p,
    color = log_fold_change
  )
) + 
  geom_point() +
  # This layer adds labels:
  geom_text_repel(
    # First, we want only a subset of data,which we can specify with the `data` argument. 
    # We can use dplyr::filter for this!
    data = dplyr::filter(gene_data, log_fold_change > 5),
    # Next, we set up the label to map to the `gene_symbol` column in our data
    aes(label = gene_symbol)
  ) + 
  theme_bw() 
```

If you prefer that all labels are the same color rather than matching the color of their points, you can specify by adding a `color` argument to `geom_text_repel`:

```{r}
ggplot(
  gene_data, 
  aes(
    x = log_fold_change,
    y = neg_log10_p,
    color = log_fold_change
  )
) + 
  geom_point() +
  # This layer adds labels:
  geom_text_repel(
    # First, we want only a subset of data,which we can specify with the `data` argument. 
    # We can use dplyr::filter for this!
    data = dplyr::filter(gene_data, log_fold_change > 5),
    # Next, we set up the label to map to the `gene_symbol` column in our data
    aes(label = gene_symbol),
    # Specify the same color (here, black) for all points
    color = "black"
  ) + 
  theme_bw() 
```

> We'll also point out this super useful Bionconductor package for volcano plots called [`EnhancedVolcano`](https://bioconductor.org/packages/release/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html), which can automatically take care of a lot of labeling and styling for you, without needing to use `ggplot2` directly.

Reviewer: I think it is worth adding a note that the plots are ggplots, but you might word it differently.
```suggestion
We'll also point out this super useful Bionconductor package for volcano plots called [`EnhancedVolcano`](https://bioconductor.org/packages/release/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html), which can automatically take care of a lot of labeling and styling for you, without needing to use `ggplot2` directly.
Handily though, the plots it creates _are_ `ggplot2` plots, so you can customize them further with the same layer syntax that you have been using!
```
